File: clients/Scala/src/main/scala/model/ColoredVertex.scala
package model

import util.StreamUtil

case class ColoredVertex(worldPos: Option[model.Vec2Float], screenOffset: model.Vec2Float, color: model.Color) {
    def writeTo(stream: java.io.OutputStream) {
        worldPos match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                value.writeTo(stream)
            }
        }
        screenOffset.writeTo(stream)
        color.writeTo(stream)
    }
}
object ColoredVertex {
    def readFrom(stream: java.io.InputStream): ColoredVertex = ColoredVertex(
        if (StreamUtil.readBoolean(stream)) Some(
            model.Vec2Float.readFrom(stream)
        ) else None
        ,
        model.Vec2Float.readFrom(stream)
        ,
        model.Color.readFrom(stream)
        )
}

File: clients/Go/model/colored_vertex.go
package model

import "io"
import . "aicup2020/stream"

type ColoredVertex struct {
    WorldPos *Vec2Float32
    ScreenOffset Vec2Float32
    Color Color
}
func NewColoredVertex(worldPos *Vec2Float32, screenOffset Vec2Float32, color Color) ColoredVertex {
    return ColoredVertex {
        WorldPos: worldPos,
        ScreenOffset: screenOffset,
        Color: color,
    }
}
func ReadColoredVertex(reader io.Reader) ColoredVertex {
    result := ColoredVertex {}
    if ReadBool(reader) {
        var WorldPosValue Vec2Float32
        WorldPosValue = ReadVec2Float32(reader)
        result.WorldPos = &WorldPosValue
    } else {
        result.WorldPos = nil
    }
    result.ScreenOffset = ReadVec2Float32(reader)
    result.Color = ReadColor(reader)
    return result
}
func (value ColoredVertex) Write(writer io.Writer) {
    if value.WorldPos == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.WorldPos).Write(writer)
    }
    value.ScreenOffset.Write(writer)
    value.Color.Write(writer)
}
